development_bank_wales/pipeline/feature_preparation/feature_engineering.py

```diff
@@ -0,0 +1,327 @@
+        Defaults to False.
+
+    Returns:
+        df (pd.DataFrame): Dataframe with upgradability score features.
+    """
+
+    for cat in cats:
+
+        total_props = df.shape[0]
+        total_props_with_rec = df.loc[(df["{}_RECOMMENDATION".format(cat)])].shape[0]
+        total_props_with_upgr = df.loc[(df["{}_EFF_DIFF".format(cat)] > 0)].shape[0]
+
+        # Energy efficiency difference or recommendation --> upgradable
+        props_w_upgr_and_rec = df.loc[
+            (
+                (df["{}_EFF_DIFF".format(cat)] > 0)
+                & (df["{}_RECOMMENDATION".format(cat)])
+            )
+        ].shape[0]
+
+        props_w_upgr_or_rec = df.loc[
+            (
+                (df["{}_EFF_DIFF".format(cat)] > 0)
+                | (df["{}_RECOMMENDATION".format(cat)])
+            )
+        ].shape[0]
+
+        df["{}_UPGRADABILITY".format(cat)] = (df["{}_EFF_DIFF".format(cat)] > 0.0) | (
+            df["{}_RECOMMENDATION".format(cat)]
+        )
+
+        if verbose:
+            print(cat)
+            print(
+                "Recommends:\t{:.2f}%".format(total_props_with_rec / total_props * 100)
+            )
+            print(
+                "Upgrades:\t{:.2f}%".format(total_props_with_upgr / total_props * 100)
+            )
+            print(
+                "Upgrade + Rec:\t{:.2f}%".format(
+                    props_w_upgr_and_rec / total_props * 100
+                )
+            )
+            print(
+                "Upgrade | Rec:\t{:.2f}%".format(
+                    props_w_upgr_or_rec / total_props * 100
+                )
+            )
+            print(
+                "Upgrade + Rec (of those with recommendation):\t{:.2f}%".format(
+                    props_w_upgr_and_rec / total_props_with_rec * 100
+                )
+            )
+            print("Coverage:\t{:.2f}%".format(props_w_upgr_or_rec / total_props * 100))
+            print(
+                "Upgr.ility:\t{:.2f}".format(df["{}_UPGRADABILITY".format(cat)].mean())
+            )
+            print()
+
+    return df
+
+
+def get_upgrade_features(df):
+    """For properties with multiple records, get information about upgrades and upgradability score.
+
+    Args:
+        df (pd.DataFrame): Property features, including ENERGY_EFF and recommendations.
+
+    Returns:
+        upgrade_df (pd.DataFrame): Dataframe with additional upgrade features.
+    """
+
+    # Clean features (to be moved to ASF core data)
+    df = extract_features_from_desc(df)
+
+    # Only consider owner-occupied properties
+    df = df.loc[df["TENURE"] == "owner-occupied"]
+
+    # Get upgrade information based on properties with multiple entries
+    latest_df = epc_data.filter_by_year(df, None, selection="latest entry")
+    first_df = epc_data.filter_by_year(df, None, selection="first entry")
+
+    upgrade_df = get_diff_in_energy_eff(first_df, latest_df, keep="first")
+    upgrade_df = compute_upgradability(upgrade_df)
```

Comment:
`compute_upgradability()` has a verbose argument but `get_upgrade_features()` doesn't.
```suggestion
    upgrade_df = compute_upgradability(upgrade_df, verbose)
```